### PR TITLE
feat: first version sourceMap

### LINKED ISSUE #3: Sourcemap support
In the latest Vite, 

> Sourcemap is likely to be incorrect: a plugin (vite-plugin-md) was used to transform files, but didn't generate a sourcemap for the transformation. Consult the plugin documentation for help

### PR DESCRIPTION
**Pull request by HomyeeKing**
close #3

### CODE REVIEW

**Review by antfu — CHANGES_REQUESTED** [review]
This is definitely good to have. But before I review your code, you can make sure your code following my ESLint rules first? Thanks!

File: src/sourcemap.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        originalLine = contentLineMap.get(line + countGE++) || originalLine
+      }
+      else {
+        countGE = 0
+        originalLine = contentLineMap.get(line) || originalLine
+      }
+
+      prevGE = line
+      const generatedLine = index + 1
+      //   console.log(line, generatedLine, originalLine)
```
Comment (HomyeeKing) [inline src/sourcemap.ts:64]: you can open this comment to  see the line map

File: package.json [hunk 2]
```diff
@@ -26,7 +26,7 @@
     "dev": "npm run build -- --watch",
     "example:dev": "cp README.md example/README.md && npm -C example run dev",
     "example:build": "cp README.md example/README.md && npm -C example run build",
-    "build": "tsup src/index.ts --dts --format cjs,esm",
+    "build": "tsup src/index.ts --dts --external @vue/compiler-sfc --format cjs,esm",
```
Comment (HomyeeKing) [inline package.json:29]: rewirth transformation with`@vue/compiler-sfc`? what do you think

**Comment by HomyeeKing** [comment]
I leave some comment and what do you think ignore transform `README.md`  by default ?

the line maps well bewteen original and generated but the error console seems not as expected 😕

File: package.json (hunk 2 above)
Comment (antfu) [reply package.json:29]: tsup should exclude depends in peerDeps already

**Comment by antfu** [comment]
> ignore transform README.md by default

Emm, No. This plugin is just a transformer and it should no have an opinion on that. If people don't want to transform README.md they should not import it.

File: package.json (hunk 2 above)
Comment (HomyeeKing) [reply package.json:29]: yeah, it's in `devDep` now
if rewrite with it, we should move it into `dep`

**Comment by HomyeeKing** [comment]
>ignore transform README.md by default

sorry ,  ignore this

File: package.json (hunk 2 above)
Comment (antfu) [reply package.json]: Umm, seems like I forgot to add it. Let's go with peerDeps, thanks!

**Comment by antfu** [comment]
I am thinking maybe we can use `magic-string` instead of `source-map`. But I am not familiar with the source map anyway, can you explain what's the trade-off between those two? I think Vite and Rollup are using `magic-string` internally, and the code of it is more readable to me as I am not familiar with source map's spec. WDYT?

File: src/markdown.ts [hunk 3]
```diff
@@ -39,26 +43,31 @@ export function createMarkdown(options: ResolvedOptions) {
...
 
     let html = markdown.render(md, {})
+    html = `\n${html}`
 
-    if (wrapperClasses)
-      html = `<div class="${wrapperClasses}">${html}</div>`
-    else
-      html = `<div>${html}</div>`
+    if (wrapperClasses) html = `<div class="${wrapperClasses}">${html}</div>`
+    else html = `<div>${html}</div>`
```
Comment (antfu) [inline src/markdown.ts:63]: Can you revert these formatting changes? Thanks

**Comment by HomyeeKing** [comment]
let me explain what I do first:

- traverse every single line of the  raw markdown code and record Map(lineContent, **originalLine** )
- tranverse every single line of  the generated code and get the current-line's **originLine** through map.get(lineContent)
- in order to get match well, both raw and generated code will store as `matter.render(lineContent)`
- so you have the **map** of the originalLine and the generatedLine now, you have the sourceMap natually


I skim the source code of `magic-string` , it seems avaliable too, but we have to rewrite the transform logic with it  since its every string manipulation methods call will record the code location

I think it worths a try with `magic-string` since it more stable and mature than mine, I will give it a try

**Comment by lhermann** [comment]
Hey here, how's the progress on sourcemap support here?

**Comment by HomyeeKing** [comment]
@lhermann  sorry for the legacy pull request
if you have interest in this issue, you can open a new PR or just work directly on this

**Comment by yankeeinlondon** [comment]
@HomyeeKing I created an updated PR which has linting errors corrected and your code incorporated into the new v0.12.x codebase. Any chance you'd be able to push this over the line?

**Comment by antfu** [comment]
Close in favor of #61

**Pull request closed without merging**